<a href="https://colab.research.google.com/github/farielshafee2018/UP-election/blob/main/copy_of_up_election_data_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
from google.colab import files
import io

This is a copy of the UP election data analysis code by Fariel Shafee

In [ ]:
districts = files.upload()

Saving UP_AC_District (2) (1)(2).csv to UP_AC_District (2) (1)(2).csv


In [ ]:
up = files.upload()

Saving UP_2022.csv to UP_2022.csv


In [ ]:
for key in districts:
  dist = pd.read_csv(io.BytesIO(districts[key]))
for key in up:
  updat= pd.read_csv(io.BytesIO(up[key]))

In [ ]:
dist.head()

AC No.           AC Name    District
0       1             Behat  Saharanpur
1       2             Nakur  Saharanpur
2       3  Saharanpur Nagar  Saharanpur
3       4        Saharanpur  Saharanpur
4       5           Deoband  Saharanpur

In [ ]:
updat.head()

ac_no      Candidate                     Party  Total Votes  % of Votes  \
0      1  Umar Ali Khan           Samajwadi Party       134513       47.81   
1      1   Naresh Saini    Bharatiya Janata Party        96633       34.34   
2      1     Rais Malik       Bahujan Samaj Party        45075       16.02   
3      1  Poonam Kamboj  Indian National Congress         1632        0.58   
4      1   Sushil Kumar           Aam Aadmi Party         1162        0.41   

   Rank  
0     1  
1     2  
2     3  
3     4  
4     5

In [ ]:
bigdat = pd.merge(updat, dist,left_on='ac_no', right_on='AC No.', how = 'inner')

In [ ]:
bigdat.head()

ac_no      Candidate                     Party  Total Votes  % of Votes  \
0      1  Umar Ali Khan           Samajwadi Party       134513       47.81   
1      1   Naresh Saini    Bharatiya Janata Party        96633       34.34   
2      1     Rais Malik       Bahujan Samaj Party        45075       16.02   
3      1  Poonam Kamboj  Indian National Congress         1632        0.58   
4      1   Sushil Kumar           Aam Aadmi Party         1162        0.41   

   Rank  AC No. AC Name    District  
0     1       1   Behat  Saharanpur  
1     2       1   Behat  Saharanpur  
2     3       1   Behat  Saharanpur  
3     4       1   Behat  Saharanpur  
4     5       1   Behat  Saharanpur

In [ ]:
bigdat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4845 entries, 0 to 4844
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   ac_no        4845 non-null   int64  
 1   Candidate    4845 non-null   object 
 2   Party        4845 non-null   object 
 3   Total Votes  4845 non-null   int64  
 4   % of Votes   4845 non-null   float64
 5   Rank         4845 non-null   int64  
 6   AC No.       4845 non-null   int64  
 7   AC Name      4845 non-null   object 
 8   District     4845 non-null   object 
dtypes: float64(1), int64(4), object(4)
memory usage: 340.8+ KB


In [ ]:
bigdat.head()

ac_no      Candidate                     Party  Total Votes  % of Votes  \
0      1  Umar Ali Khan           Samajwadi Party       134513       47.81   
1      1   Naresh Saini    Bharatiya Janata Party        96633       34.34   
2      1     Rais Malik       Bahujan Samaj Party        45075       16.02   
3      1  Poonam Kamboj  Indian National Congress         1632        0.58   
4      1   Sushil Kumar           Aam Aadmi Party         1162        0.41   

   Rank  AC No. AC Name    District  
0     1       1   Behat  Saharanpur  
1     2       1   Behat  Saharanpur  
2     3       1   Behat  Saharanpur  
3     4       1   Behat  Saharanpur  
4     5       1   Behat  Saharanpur

In [ ]:
winners=bigdat[bigdat["Rank"]==1]
winners.head()

ac_no         Candidate                   Party  Total Votes  % of Votes  \
0       1     Umar Ali Khan         Samajwadi Party       134513       47.81   
9       2  MUKESH CHOUDHARY  Bharatiya Janata Party       104114       38.47   
21      3     RAJEEV GUMBER  Bharatiya Janata Party       143195       48.85   
34      4        Ashu Malik         Samajwadi Party       107007       41.18   
49      5           BRIJESH  Bharatiya Janata Party        93890       38.77   

    Rank  AC No.           AC Name    District  
0      1       1             Behat  Saharanpur  
9      1       2             Nakur  Saharanpur  
21     1       3  Saharanpur Nagar  Saharanpur  
34     1       4        Saharanpur  Saharanpur  
49     1       5           Deoband  Saharanpur

In [ ]:
winners["Party"].value_counts()


Party
Bharatiya Janata Party                  255
Samajwadi Party                         111
Apna Dal (Soneylal)                      12
Rashtriya Lok Dal                         8
Suheldev Bharatiya Samaj Party            6
Nirbal Indian Shoshit Hamara Aam Dal      6
Indian National Congress                  2
Jansatta Dal Loktantrik                   2
Bahujan Samaj Party                       1
Name: count, dtype: int64

### 1 Which party will form the govt?

### answer:  BJP

### 2. Show partywise percentage of top ten parties

In [ ]:
bigdat['partytot']=bigdat.groupby('Party')['Total Votes'].transform('sum')

In [ ]:
bigdat.head(10)

ac_no         Candidate                     Party  Total Votes  % of Votes  \
0      1     Umar Ali Khan           Samajwadi Party       134513       47.81   
1      1      Naresh Saini    Bharatiya Janata Party        96633       34.34   
2      1        Rais Malik       Bahujan Samaj Party        45075       16.02   
3      1     Poonam Kamboj  Indian National Congress         1632        0.58   
4      1      Sushil Kumar           Aam Aadmi Party         1162        0.41   
5      1              NOTA         None of the Above         1117        0.40   
6      1   Dharampal Singh               Independent          560        0.20   
7      1          Ali Khan               Independent          409        0.15   
8      1         Mo. IKram               Independent          272        0.10   
9      2  MUKESH CHOUDHARY    Bharatiya Janata Party       104114       38.47   

   Rank  AC No. AC Name    District  partytot  
0     1       1   Behat  Saharanpur  29543934  
1     2       1   Behat  Saharanpur  38051721  
2     3       1   Behat  Saharanpur  11873137  
3     4       1   Behat  Saharanpur   2151234  
4     5       1   Behat  Saharanpur    347192  
5     6       1   Behat  Saharanpur    637304  
6     7       1   Behat  Saharanpur   1019576  
7     8       1   Behat  Saharanpur   1019576  
8     9       1   Behat  Saharanpur   1019576  
9     1       2   Nakur  Saharanpur  38051721

In [ ]:
sumvote=bigdat['Total Votes'].sum()
sumvote

np.int64(92162896)

In [ ]:
bigdat['Party Percentage'] = (bigdat['partytot']/sumvote)*100
bigdat.head()

ac_no      Candidate                     Party  Total Votes  % of Votes  \
0      1  Umar Ali Khan           Samajwadi Party       134513       47.81   
1      1   Naresh Saini    Bharatiya Janata Party        96633       34.34   
2      1     Rais Malik       Bahujan Samaj Party        45075       16.02   
3      1  Poonam Kamboj  Indian National Congress         1632        0.58   
4      1   Sushil Kumar           Aam Aadmi Party         1162        0.41   

   Rank  AC No. AC Name    District  partytot  Party Percentage  
0     1       1   Behat  Saharanpur  29543934         32.056213  
1     2       1   Behat  Saharanpur  38051721         41.287462  
2     3       1   Behat  Saharanpur  11873137         12.882773  
3     4       1   Behat  Saharanpur   2151234          2.334165  
4     5       1   Behat  Saharanpur    347192          0.376716

In [ ]:
top_ten=bigdat[['Party','Party Percentage']].drop_duplicates().sort_values(by='Party Percentage', ascending = False)
top_ten.head(10)

Party  Party Percentage
1                   Bharatiya Janata Party         41.287462
0                          Samajwadi Party         32.056213
2                      Bahujan Samaj Party         12.882773
62                       Rashtriya Lok Dal          2.853825
3                 Indian National Congress          2.334165
395                    Apna Dal (Soneylal)          1.620154
1760        Suheldev Bharatiya Samaj Party          1.359468
6                              Independent          1.106276
2538  Nirbal Indian Shoshit Hamara Aam Dal          0.912063
5                        None of the Above          0.691497

### #3 How many constituencies are there where the difference between Bharatiya Janata Party and Samajwadi Party is less than 5000

In [ ]:
# Filter the dataframe for 'Bharatiya Janata Party' and 'Samajwadi Party'
bjp_sp_votes = bigdat[bigdat['Party'].isin(['Bharatiya Janata Party', 'Samajwadi Party'])]

# Pivot the dataframe to have parties as columns and total votes as values
pivot_votes = bjp_sp_votes.pivot_table(index=['ac_no', 'AC Name', 'District'], columns='Party', values='Total Votes', aggfunc='sum').reset_index()

# Calculate the absolute difference in votes between the two parties
pivot_votes['Vote Difference'] = abs(pivot_votes['Bharatiya Janata Party'] - pivot_votes['Samajwadi Party'])

# Filter for constituencies where the vote difference is less than 5000
close_constituencies = pivot_votes[pivot_votes['Vote Difference'] < 5000]

# Display the result
print("Constituencies where the vote difference between Bharatiya Janata Party and Samajwadi Party is less than 5000:")
display(close_constituencies)

# count total
close_constituencies['ac_no'].count()

Constituencies where the vote difference between Bharatiya Janata Party and Samajwadi Party is less than 5000:


Party  ac_no          AC Name            District  Bharatiya Janata Party  \
1          2            Nakur          Saharanpur                104114.0   
19        20          Dhampur              Bijnor                 81791.0   
22        23         Chandpur              Bijnor                 90288.0   
27        28  Moradabad Nagar           Moradabad                148384.0   
35        36         Bilaspur              Rampur                101998.0   
45        46          Kithore              Meerut                104924.0   
95        96          Jasrana           Firozabad                107453.0   
101      102         Patiyali                Etah                 87957.0   
102      103          Aliganj                Etah                102873.0   
105      106          Jalesar                Etah                 91339.0   
107      108         Bhongaon            Mainpuri                 97208.0   
111      112          Bisauli              Budaun                108735.0   
117      118           Baheri            Bareilly                120790.0   
121      122         Faridpur            Bareilly                 92070.0   
130      131            Katra        Shahjahanpur                 77800.0   
131      132        Jalalabad        Shahjahanpur                 99609.0   
143      144         Mohammdi               Kheri                 99377.0   
145      146          Sitapur             Sitapur                 99349.0   
180      181            Salon          Rae Bareli                 87715.0   
181      182           Sareni          Rae Bareli                 62359.0   
186      187           Isauli           Sultanpur                 69360.0   
187      188        Sultanpur           Sultanpur                 92715.0   
195      196       Chhibramau             Kannauj                124773.0   
196      197            Tirwa             Kannauj                106089.0   
199      200           Etawah              Etawah                 98150.0   
202      203        Dibiyapur             Auraiya                 80392.0   
249      250         Raniganj          Pratapgarh                 72934.0   
255      256          Phulpur           Allahabad                103557.0   
258      259             Meja           Allahabad                 75116.0   
265      266            Kursi           Barabanki                118720.0   
266      267        Ram Nagar           Barabanki                 98538.0   
268      269          Zaidpur           Barabanki                110576.0   
285      286         Bahraich            Bahraich                107628.0   
289      290        Shrawasti           Shrawasti                 98640.0   
303      305             Itwa      Siddharthnagar                 62591.0   
304      306     Domariyaganj      Siddharthnagar                 84327.0   
308      310      Basti Sadar               Basti                 84250.0   
350      353         Madhuban                 Mau                 79032.0   
361      364         Badlapur             Jaunpur                 82391.0   
372      375         Ghazipur            Ghazipur                 90780.0   
388      392          Bhadohi  S.R.Nagar(Bhadohi)                 95853.0   
390      394            Aurai  S.R.Nagar(Bhadohi)                 93691.0   

Party  Samajwadi Party  Vote Difference  
1             103799.0            315.0  
19             81588.0            203.0  
22             90522.0            234.0  
27            147602.0            782.0  
35            101691.0            307.0  
45            107104.0           2180.0  
95            108289.0            836.0  
101            91958.0           4001.0  
102            99063.0           3810.0  
105            86898.0           4441.0  
107            92441.0           4767.0  
111           110569.0           1834.0  
117           124145.0           3355.0  
121            89149.0           2921.0  
130            77443.0            357.0  
131            9

np.int64(42)

4. In which of the constituencies described in 3 did BJP lose?

In [ ]:
# first find actual difference


pivot_votes['actual difference'] = pivot_votes['Bharatiya Janata Party'] - pivot_votes['Samajwadi Party']
# find total close losses


total_close_loss=pivot_votes[(pivot_votes['actual difference']<0) & (pivot_votes['actual difference']>-5000)]["ac_no"].count()
total_close_loss

np.int64(18)

5. How many independent candidates won the election?

In [ ]:
bigdat.head(20)

ac_no               Candidate                                   Party  \
0       1           Umar Ali Khan                         Samajwadi Party   
1       1            Naresh Saini                  Bharatiya Janata Party   
2       1              Rais Malik                     Bahujan Samaj Party   
3       1           Poonam Kamboj                Indian National Congress   
4       1            Sushil Kumar                         Aam Aadmi Party   
5       1                    NOTA                       None of the Above   
6       1         Dharampal Singh                             Independent   
7       1                Ali Khan                             Independent   
8       1               Mo. IKram                             Independent   
9       2        MUKESH CHOUDHARY                  Bharatiya Janata Party   
10      2  DR. DHARAM SINGH SAINI                         Samajwadi Party   
11      2              SAHIL KHAN                     Bahujan Samaj Party   
12      2                 RIZWANA  All India Majlis-E-Ittehadul Muslimeen   
13      2          RANDHEER SINGH                Indian National Congress   
14      2                    NOTA                       None of the Above   
15      2              SONU KUMAR          Aazad Samaj Party (Kanshi Ram)   
16      2              ATUL KUMAR                       Inqalab Vikas Dal   
17      2             NITIN KUMAR                  Jai Maha Bharath Party   
18      2            ASHISH TYAGI                         Aam Aadmi Party   
19      2        DEVENDRA CHAUHAN                       Aamjan Ekta Party   

    Total Votes  % of Votes  Rank  AC No. AC Name    District  partytot  \
0        134513       47.81     1       1   Behat  Saharanpur  29543934   
1         96633       34.34     2       1   Behat  Saharanpur  38051721   
2         45075       16.02     3       1   Behat  Saharanpur  11873137   
3          1632        0.58     4       1   Behat  Saharanpur   2151234   
4          1162        0.41     5       1   Behat  Saharanpur    347192   
5          1117        0.40     6       1   Behat  Saharanpur    637304   
6           560        0.20     7       1   Behat  Saharanpur   1019576   
7           409        0.15     8       1   Behat  Saharanpur   1019576   
8           272        0.10     9       1   Behat  Saharanpur   1019576   
9        104114       38.47     1       2   Nakur  Saharanpur  38051721   
10       103799       38.35     2       2   Nakur  Saharanpur  29543934   
11        55112       20.36     3       2   Nakur  Saharanpur  11873137   
12         3593        1.33     4       2   Nakur  Saharanpur    450929   
13         1354        0.50     5       2   Nakur  Saharanpur   2151234   
14          710        0.26     6       2   Nakur  Saharanpur    637304   
15          616        0.23     7       2   Nakur  Saharanpur    119564   
16          420        0.16     8       2   Nakur  Saharanpur       420   
17          267        0.10     9       2   Nakur  Saharanpur       267   
18          240        0.09    10       2   Nakur  Saharanpur    347192   
19          218        0.08    11       2   Nakur  Saharanpur       409   

    Party Percentage  
0          32.056213  
1          41.287462  
2          12.882773  
3           2.334165  
4           0.376716  
5           0.691497  
6           1.106276  
7           1.106276  
8           1.106276  
9          41.287462  
10         32.056213  
11         12.882773  
12          0.489274  
13          2.334165  
14          0.691497  
15          0.129731  
16          0.000456  
17          0.000290  
18          0.376716  
19          0.000444

In [ ]:


independent_win = bigdat[(bigdat['Party'] =='Independent') & (bigdat['Rank'] == 1)]
independent_win.head()
independent_win['ac_no'].count()


np.int64(0)

5. Which are the top five districts with the highest number of voters?

In [ ]:
# find total votes in each district

dist_votes = bigdat.groupby('District')['Total Votes'].sum().reset_index()
dist_votes.head()


District  Total Votes
0            Agra      2116833
1         Aligarh      1728326
2       Allahabad      2504317
3  Ambedkar Nagar      1148997
4         Auraiya       624541

In [ ]:
# find top five

top_five_dists= dist_votes.groupby('District')['Total Votes'].sum().sort_values(ascending=False).head(5)
top_five_dists

District
Allahabad    2504317
Moradabad    2366697
Lucknow      2318278
Ghaziabad    2302903
Agra         2116833
Name: Total Votes, dtype: int64

7. Which party was the bigges competitior for teh Bharatiya Janata Party?


In [ ]:
party_votes = bigdat.groupby('Party')['Total Votes'].sum().reset_index()
party_votes.head()
sorted = party_votes.sort_values(by='Total Votes', ascending=False)
sorted.head()
second_party = sorted.iloc[1]['Party']
second_party

'Samajwadi Party'

8. What was the minimum votes difference the Bharatiya Janata party candidate lost with and with which party candidate?

In [ ]:
# find all losing BJP members

losing_bjp = bigdat[(bigdat['Party'] == 'Bharatiya Janata Party') & (bigdat['Rank'] == 2)]
losing_bjp.head()

# find the difference of votes between all Bharatiya Janata party members who lost and the candidate who won
# first find the consituency

const = losing_bjp['ac_no'].unique()

# find winners

relev_subset = bigdat[(bigdat['ac_no'].isin(const))]

winner = relev_subset[relev_subset["Rank"]==1]

# merge winner and bjp

#merged_set = pd.merge(losing_bjp, winner, on = "ac_no", how = "inner")
#merged_set.head()

# subset including just those constituencies merging winner and bjp

bjp_loss_subset = pd.merge(losing_bjp, winner[['ac_no', 'Total Votes', 'Candidate', 'Party']], on = "ac_no", suffixes = ("_bjp", "_winner"))

bjp_loss_subset.head()


#find difference and min

diff = abs(bjp_loss_subset['Total Votes_bjp'] - bjp_loss_subset['Total Votes_winner'])
min_diff = diff.min()
print(f'min diff is {min_diff}')





min diff is 234


#9 If you want to make one recommendation to BJP to win more seats, what would it be?



I would ask BJP to focus on the close seats where their main competitor, Samajwadi Party, won by narrow margins.  The minimum disserence is just 234 and it could easily swing in the other direction.  BJP lost in 18 of the 42 closely contested seats.  They should make an effort to increase their share of votes in all these consituencies to increase numbers of wins and also to maintain the seats where they won only marginally.

10 Which district has the maximum number of consituencies?

In [ ]:
const = bigdat.groupby('District')['ac_no'].nunique().reset_index()
const.head()
const.sort_values(by='ac_no', ascending = False).head()

District  ac_no
2      Allahabad     12
5       Azamgarh     10
39  Kanpur Nagar     10
0           Agra      9
12      Bareilly      9

11. Which are the top five districts of BJP?

In [ ]:
# make subset where BJP won

bjpsub = bigdat[(bigdat['Party'] == 'Bharatiya Janata Party') & (bigdat['Rank'] == 1)]
bjpsub.head()

ac_no           Candidate                   Party  Total Votes  \
9       2    MUKESH CHOUDHARY  Bharatiya Janata Party       104114   
21      3       RAJEEV GUMBER  Bharatiya Janata Party       143195   
49      5             BRIJESH  Bharatiya Janata Party        93890   
60      6  Devender Kumar Nim  Bharatiya Janata Party        89109   
70      7         KIRAT SINGH  Bharatiya Janata Party       116582   

    % of Votes  Rank  AC No.           AC Name    District  partytot  \
9        38.47     1       2             Nakur  Saharanpur  38051721   
21       48.85     1       3  Saharanpur Nagar  Saharanpur  38051721   
49       38.77     1       5           Deoband  Saharanpur  38051721   
60       38.37     1       6  Rampur Maniharan  Saharanpur  38051721   
70       43.08     1       7            Gangoh  Saharanpur  38051721   

    Party Percentage  
9          41.287462  
21         41.287462  
49         41.287462  
60         41.287462  
70         41.287462

In [ ]:
# now find which districts are represented the most in BJP wins

bjpsub.groupby('District')['ac_no'].count().sort_values(ascending = False).head(20)

District
Agra            9
Gorakhpur       9
Ghaziabad       8
Hardoi          8
Kheri           8
Sitapur         8
Allahabad       7
Deoria          7
Aligarh         7
Gonda           7
Varanasi        7
Bulandshahar    7
Bareilly        7
Lucknow         7
Etah            6
Shahjahanpur    6
Unnao           6
Kanpur Nagar    6
Kushinagar      6
Saharanpur      5
Name: ac_no, dtype: int64

12  In which constituency did the maximum number of candidates contest the election?

In [ ]:
# we need to group by constituencies and count the number of candidates and then sort in descending order


grouped_const = bigdat.groupby('ac_no')['Candidate'].count().reset_index()
grouped_const.head()
grouped_const.sort_values(by = 'Candidate', ascending = False).head()



ac_no  Candidate
365    366         26
256    257         26
253    254         22
374    375         20
246    247         20

13. How many parties lost all the elections they contested (except independents)?

In [ ]:
# first remove all independents

bigdat_no_indep = bigdat[bigdat['Party'] != 'Independent']
bigdat_no_indep.head()

ac_no      Candidate                     Party  Total Votes  % of Votes  \
0      1  Umar Ali Khan           Samajwadi Party       134513       47.81   
1      1   Naresh Saini    Bharatiya Janata Party        96633       34.34   
2      1     Rais Malik       Bahujan Samaj Party        45075       16.02   
3      1  Poonam Kamboj  Indian National Congress         1632        0.58   
4      1   Sushil Kumar           Aam Aadmi Party         1162        0.41   

   Rank  AC No. AC Name    District  partytot  Party Percentage  
0     1       1   Behat  Saharanpur  29543934         32.056213  
1     2       1   Behat  Saharanpur  38051721         41.287462  
2     3       1   Behat  Saharanpur  11873137         12.882773  
3     4       1   Behat  Saharanpur   2151234          2.334165  
4     5       1   Behat  Saharanpur    347192          0.376716

In [ ]:
# now in the big_dat_no_indep dataframe, find parties that have no Rank 1 anywhere.  The other way to do it is to create a subdataframe removing the rank 1's and lost all parties there and then compare with all parties in the bigdat_no_indep to see which party is missing in the new frame that has no rank 1

no_rank_1 = bigdat_no_indep[bigdat_no_indep['Rank'] != 1]
no_rank_1.head()




ac_no      Candidate                     Party  Total Votes  % of Votes  \
1      1   Naresh Saini    Bharatiya Janata Party        96633       34.34   
2      1     Rais Malik       Bahujan Samaj Party        45075       16.02   
3      1  Poonam Kamboj  Indian National Congress         1632        0.58   
4      1   Sushil Kumar           Aam Aadmi Party         1162        0.41   
5      1           NOTA         None of the Above         1117        0.40   

   Rank  AC No. AC Name    District  partytot  Party Percentage  
1     2       1   Behat  Saharanpur  38051721         41.287462  
2     3       1   Behat  Saharanpur  11873137         12.882773  
3     4       1   Behat  Saharanpur   2151234          2.334165  
4     5       1   Behat  Saharanpur    347192          0.376716  
5     6       1   Behat  Saharanpur    637304          0.691497

In [ ]:
# list all party names in no_rank_1

losing_parties = no_rank_1['Party'].unique()
losing_parties



array(['Bharatiya Janata Party', 'Bahujan Samaj Party',
       'Indian National Congress', 'Aam Aadmi Party', 'None of the Above',
       'Samajwadi Party', 'All India Majlis-E-Ittehadul Muslimeen',
       'Aazad Samaj Party (Kanshi Ram)', 'Inqalab Vikas Dal',
       'Jai Maha Bharath Party', 'Aamjan Ekta Party', 'Lok Dal',
       'Akhand Bharat Vikas Party', 'Bahujan Mukti Party',
       'Sanyukt Vikas Party', 'Log Party',
       'Bharatrashtra Democratic Party', 'Jansatta Party',
       'Rashtriya Lok Dal', 'India Janshakti Party',
       'SOCIAL DEMOCRATIC PARTY OF INDIA', 'Sarvjan Lok Shakti Party',
       'Jai Samta Party', 'Rashtriya Samaj Paksha',
       'Majdoor Kisan Union Party', 'Bharat Lok Sewak Party',
       'Bharatiya Sarvodaya Kranti Party', 'Shiv Sena',
       'Pichhra Samaj Party', 'Rashtra Nirman Party',
       'Rashtriya Samaj Dal (R)', 'Gareeb Kranti Party',
       'Rashtriya mahan Gantantra Party', 'Ambedkar Samaj Party',
       'Peoples Party of India (Democratic

In [ ]:
rty

In [ ]:
# now list all parties that have some rank 1

some_rank_1 = bigdat_no_indep[bigdat_no_indep['Rank'] == 1]['Party'].unique()
some_rank_1

array(['Samajwadi Party', 'Bharatiya Janata Party', 'Rashtriya Lok Dal',
       'Apna Dal (Soneylal)', 'Indian National Congress',
       'Jansatta Dal Loktantrik', 'Suheldev Bharatiya Samaj Party',
       'Nirbal Indian Shoshit Hamara Aam Dal', 'Bahujan Samaj Party'],
      dtype=object)

In [ ]:
# now find which parties are in losing_parties but not in some_rank_1

no_winners = [party for party in losing_parties if party not in some_rank_1]
no_winners

['Aam Aadmi Party',
 'None of the Above',
 'All India Majlis-E-Ittehadul Muslimeen',
 'Aazad Samaj Party (Kanshi Ram)',
 'Inqalab Vikas Dal',
 'Jai Maha Bharath Party',
 'Aamjan Ekta Party',
 'Lok Dal',
 'Akhand Bharat Vikas Party',
 'Bahujan Mukti Party',
 'Sanyukt Vikas Party',
 'Log Party',
 'Bharatrashtra Democratic Party',
 'Jansatta Party',
 'India Janshakti Party',
 'SOCIAL DEMOCRATIC PARTY OF INDIA',
 'Sarvjan Lok Shakti Party',
 'Jai Samta Party',
 'Rashtriya Samaj Paksha',
 'Majdoor Kisan Union Party',
 'Bharat Lok Sewak Party',
 'Bharatiya Sarvodaya Kranti Party',
 'Shiv Sena',
 'Pichhra Samaj Party',
 'Rashtra Nirman Party',
 'Rashtriya Samaj Dal (R)',
 'Gareeb Kranti Party',
 'Rashtriya mahan Gantantra Party',
 'Ambedkar Samaj Party',
 'Peoples Party of India (Democratic)',
 'Loktanter Suraksha Party',
 'Bhartiya Jan Samman Party',
 'Jan Adhikar Party',
 'Bhartiya Damitjan Party',
 'Rashtrawadi Janlok Party (Satya)',
 'Samajwadi Andolan Party',
 'Peace Party',
 'Rashtriya 